### PR TITLE
Fix seq2seq collator padding

### LINKED ISSUE #30521: [BUG] DataCollatorForSeq2Seq with PaddingStrategy.MAX_LENGTH may not pad labels
It seems that when padding, if the MAX_LENGTH policy is set, the same padding is not performed on the label.

test case below：
```python
from transformers import DataCollatorForSeq2Seq,
from transformers.utils import PaddingStrategy
inputs=[{'input_ids': [151644, 8948, 198],'attention_mask': [1, 1, 1],'labels': [-100, -100, -100]},
 {'input_ids': [151644, 8948, 198, 2610],'attention_mask': [1, 1, 1, 1],'labels': [-100, -100, -100, -100]},
 {'input_ids': [151644, 8948, 198, 2610, 525], 'attention_mask': [1, 1, 1, 1, 1],'labels': [-100, -100, -100, -100, -100]}]
data_collator = DataCollatorForSeq2Seq(
        tokenizer=tokenizer,
        padding=PaddingStrategy.MAX_LENGTH,
        max_length=10,
    )
res=data_collator(inputs)

print(res['input_ids'].shape,res['labels'].shape)
```

results:
torch.Size([3, 10]) torch.Size([3, 5])

expected results:
torch.Size([3, 10]) torch.Size([3, 10])

Should the following code handle the pad length of the label according to different strategies?
https://github.com/huggingface/transformers/blob/73014b561d5f88d728e46a57d346f516fefe3f2d/src/transformers/data/data_collator.py#L592

### PR DESCRIPTION
**Pull request by vasqu**
# What does this PR do?
`DataCollatorForSeq2Seq` currently only supports the `longest` and silently the `no_padding` strategies. This PR adds some checks to ensure the padding strategy is respected if valid (e.g. `max_length` without a given max length defaults back to `longest`).

Furthermore, added tests for the collator in the style of the `DataCollatorForTokenClassification` for all kinds of supported datatypes (pt, tf, np). 

Something to note: The seq2seq collator causes side effects on the features' labels that are passed. I don't think it's a big deal but thought I'd clarify why I produce features twice within a single test.

Fixes #30521 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@Rocketknight1 @amyeroberts

### CODE REVIEW

**Review by Rocketknight1 — APPROVED** [review]
Code changes LGTM, and I appreciate the extensive testing! cc @amyeroberts for core maintainer review.

The point you made about side-effects on the input is interesting, though - I agree that it's unlikely to be critical (because people rarely reuse the pre-collator inputs after calling the collator), but it's still suboptimal. Maybe we could consider having our collators make copies of the inputs instead so they don't overwrite them, though this would definitely be a task for a separate PR.

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing and for all of these tests - beautiful!

**Comment by amyeroberts** [comment]
> The point you made about side-effects on the input is interesting, though - I agree that it's unlikely to be critical (because people rarely reuse the pre-collator inputs after calling the collator), but it's still suboptimal. Maybe we could consider having our collators make copies of the inputs instead so they don't overwrite them, though this would definitely be a task for a separate PR.

@Rocketknight1 @vasqu Very much aligned on this - side-effects should definitely be avoided, and it's a bigger piece of work for a follow-up PR.

**Comment by vasqu** [comment]
@amyeroberts @Rocketknight1 I'm not sure which collators affect which inputs via side-effects so I'd suggest writing a mutability test which assures this wanted behaviour. Afaik, it currently only affects the labels when using the seq2seq collator but maybe I'll catch some others. Got some time now, will be looking into it in a bit.

**Comment by amyeroberts** [comment]
@vasqu Awesome ❤️  looking forward to seeing the PR!
